docs/developer-docs/security/security-best-practices/data-integrity-and-authenticity.mdx

```diff
@@ -8,19 +8,334 @@ import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";
+| Query call is analogous to _READ_                                                                      | Update call is analogous to _WRITE_                                                                                                                       |
+| The query call is executed on a single replica node chosen on random in a subnet by the boundary nodes | On an update call, the request is replicated among all the nodes in the subnet, goes through consensus and the response is threshold signed by the subnet |
+| **Very fast response time**                                                                            | Relatively slow and more expensive (depending on the size of the subnet).                                                                                 |
+| Relatively less secure since there is no integrity protection. The responding node maybe be malicious  | **Secure response, since it’s certified by the subnet and guarantees integrity**                                                                          |
 
-### Recommendation
 
--   All security-relevant query response data that needs authenticity guarantees should be certified by ICP using certified variables. This determination needs to be assessed for each dapp. Consider using existing data structures such as [certified-map](https://github.com/dfinity/cdk-rs/tree/main/library/ic-certified-map). The data certification must be validated in the frontend.
+Queries have fast response time but lack integrity guarantees, which makes them unfit for security critical applications. To obtain query like response times and update like certified responses, developers could make use of `certified_data` in the state tree to certify the response served in a query. Alternatively, these calls could be issued as update calls by the caller (e.g. in agent-js), but that impacts performance by taking a few seconds. Some examples of Certified Variables are asset certification in [Internet Identity](https://github.com/dfinity/internet-identity/blob/b29a6f68bbe5a49d048e12bc7a3263a9f43d080b/src/internet_identity/src/main.rs#L775-L808), [NNS dapp](https://github.com/dfinity/nns-dapp/blob/372c3562127d70c2fde059bc9c268e8ae858583e/rs/src/assets.rs#L121-L145), or the [canister signature implementation in Internet Identity](https://github.com/dfinity/internet-identity/blob/main/src/internet_identity/src/signature_map.rs).
+
+:::tip
+ICP also provides Replica signed queries, where query responses are signed by the answering replica node but however it doesn't have the same security guarantees as `update` calls. Replica signed queries are enabled by default on both `agent-rs` and `agent-js`. You can read more about the feature [here](https://internetcomputer.org/blog/features/replica-signed-queries)
+:::
+
+### What's Certified data?
```

Comment:
I would elaborate on this a bit more. Maybe using a diagram that includes the data struct, the hash, the signature/certificate and where it is updated/created (update), retrieved (query) and validated (client).

I'm also missing the drawbacks/caveats of certified data. For example, this doesn't work if your query call returns a calculation result since only update calls can update the certificate. The workaround it to give all the certified source data to the client which can check the certificate and make the same calculations or to have certain pre calculated values certified in update calls.

Reply:
I think this is already explained in Certified Variables section and I have referenced the link.

Reply:
Maybe [this](https://f7i7o-aaaaa-aaaam-abtba-cai.icp0.io/how-it-works/response-certification/#certified-variables) is also a good resource. There is even a 30min [YouTube video](https://www.youtube.com/watch?v=3mZHEfICi_U) we could link.

Reply:
Added a section with a high level explanation of certified variables.